### DISCUSSION TITLE
Allow request body for model only if a property has a certain type?

### DISCUSSION THREAD

**Discussion by danielniccoli** [fastapi/fastapi#10120] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel


class IntComponent(BaseModel):
    value: int

class StringComponent(BaseModel):
    value: str

class MyModel(BaseModel):
    component: IntComponent | StringComponent


app = FastAPI()


@app.post("/items/")
async def create_item(item: MyModel):
    return item
```

### Description

How can I define that for a particular path like `async def create_item(item: MyModel):` I accept only an instance of `MyModel` **if** the `my_component` property is of type `MyComponent1`. Validation should fail, if the `my_component` object type is an instance of `MyComponent2`.

With the described code, a request body of `{ "component": { "value": 1 } }` should pass, but `{ "component": { "value": "one" } }` must fail. How do I have to change the line `async def create_item(item: MyModel):` to achieve this?

### Operating System

Linux

### FastAPI Version

0.101.1

### Pydantic Version

2.1.1

### Python Version

3.10.11

**Comment by ebottos94** [comment, marked as answer]
Simply creating another Pydantic model. Now you are allowing that "value" could be string or integer...

**Reply by danielniccoli** [reply]
I want to achieve the goal without creating another model. It must work with the composite model.

**Reply by Kludex** [reply]
You can't.

**Reply by danielniccoli** [reply]
@Kludex is that a restriction on FastAPI? Or Pydantic? I'd like to open a feature request to support composited objects.

**Reply by Kludex** [reply]
I can't see the need for it. How do you see that feature from the point of view of a user?

**Reply by danielniccoli** [reply]
@Kludex Is this the correct place to discuss this? Is this a restriction with FastAPI, or does it come from Pydantic?

**Reply by Kludex** [reply]
>  Is this the correct place to discuss this?

Yes.

> Is this a restriction with FastAPI, or does it come from Pydantic?

I don't think it's a restriction... If you can reply my question, I can tell you the next steps - if any.

**Reply by danielniccoli** [reply]
> from the point of view of a user

I'm not sure what you mean by user. If you mean the person opening the web browser, then no, this is not about that. It's about the choice of inheritance vs composition in the code.

**Reply by Kludex** [reply]
I mean you, the developer.

**Reply by danielniccoli** [reply]
It's about the choice of inheritance vs composition in the code.
